### PR TITLE
Use an extra shims header to remove _silgen_name from Dispatch.

### PR DESCRIPTION
**Pull request by jrose-apple**
Follow-up to #5309 for Dispatch.

We still have a bunch of redeclarations of Dispatch functions to avoid the automatic bridging of dispatch_data_t and dispatch_block_t, but mostly this is a vast reduction in complexity (and increase in safety).

I think there are still bugs, so I'm not going to call any of the Dispatch folks to review just yet.

After this gets merged, we'll only be using _silgen_name outside the core stdlib in Foundation and "Platform" (Darwin or Glibc), plus a number of things that don't actually get shipped (benchmarks and StdlibUnittest).

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please ASan test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
Filed rdar://problem/29340689 for the odd compiler-rt failure.

@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
[ASan test happening here](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-ASAN-test/42/).

**Comment by jrose-apple** [comment]
Ready for review! @mwwa, @das?

@swift-ci Please test macOS
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
Ready for review! @mwwa, @das?

@swift-ci Please test macOS
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by das** [comment]
do we need an equivalent change to swift-corelibs-libdispatch for the Linux overlay ? I see _silgen_name in src/swift/ files

File: stdlib/public/SwiftShims/DispatchOverlayShims.h [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+_swift_dispatch_data_destructor_free(void) {
+  return _dispatch_data_destructor_free;
+}
+
+static inline __swift_shims_dispatch_block_t
+_swift_dispatch_data_destructor_munmap(void) {
+  return _dispatch_data_destructor_munmap;
+}
+
+#define SWIFT_DISPATCH_SOURCE(t)                                               \
```
Comment (das) [inline stdlib/public/SwiftShims/DispatchOverlayShims.h]: SWIFT_DISPATCH_SOURCE_TYPE ?

File: stdlib/public/SwiftShims/DispatchOverlayShims.h [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    __swift_shims_dispatch_block_t block) {
+  dispatch_sync(queue, block);
+}
+
+static inline void _swift_dispatch_barrier_async(
+    dispatch_queue_t queue,
+    __swift_shims_dispatch_block_t block) {
+  dispatch_barrier_async(queue, block);
+}
+
```
Comment (das) [inline stdlib/public/SwiftShims/DispatchOverlayShims.h]: more a question for @mwwa : why don't we have _swift_dispatch_barrier_sync here (i.e. why can _syncBarrier use __dispatch_barrier_sync ) ?

**Comment by das** [comment]
lgtm but @mwwa should have a look too since he is the one that originally wrote this code

File: stdlib/public/SwiftShims/DispatchOverlayShims.h (hunk 1 above)
Comment (jrose-apple) [inline stdlib/public/SwiftShims/DispatchOverlayShims.h]: Ah, good call, will change.

**Comment by jrose-apple** [comment]
I forgot that the Linux overlay would have _silgen_name in it too. I don't think anything there will break, but if they depend on DispatchShims.mm they might. Let's find out.

@swift-ci Please test

File: stdlib/public/SwiftShims/DispatchOverlayShims.h (hunk 2 above)
Comment (mwwa) [inline stdlib/public/SwiftShims/DispatchOverlayShims.h]: We don't have a `sync<T>` call that takes a `DispatchWorkItem`. The reason we have to have these shims is to preserve the identity of `DispatchWorkItem/dispatch_block_t` such that the runtime doesn't wrap them in thunk closures. The barrier case in `sync<T>` is only ever fed closures anyway, so we don't need a shim here.

**Comment by mwwa** [comment]
Same, LGTM.

**Comment by jrose-apple** [comment]
All tests passed, so I'll update the name for SWIFT_DISPATCH_SOURCE_TYPE, commit, and then do swift-corelibs-dispatch separately.

**Comment by jrose-apple** [comment]
Full tests passed already, so just running smoke tests this time.

@swift-ci Please smoke test

**Pull request merged**